src/test/java/com/shapesecurity/shift/es2016/test262/Test262.java

```diff
@@ -0,0 +1,196 @@
+				return null;
+			}
+			Map<String, Object> negative = (Map<String, Object>) rawNegative;
+			String phase = (String) negative.get("phase");
+			if (phase == null) {
+				return null;
+			}
+			switch (phase) {
+				case "parse":
+					negativeEnum = Test262Info.Test262Negative.PARSE;
+					break;
+				case "early":
+					negativeEnum = Test262Info.Test262Negative.EARLY;
+					break;
+				default:
+					negativeEnum = Test262Info.Test262Negative.EXECUTE;
+					break;
+			}
+		}
+		Object rawFlags = parsedYaml.get("flags");
+		boolean noStrict = false;
+		boolean onlyStrict = false;
+		boolean async = false;
+		boolean module = false;
+		if (rawFlags != null) {
+			ArrayList<String> flags = (ArrayList<String>) rawFlags;
+			for (String flag : flags) {
+				switch (flag) {
+					case "noStrict":
+						noStrict = true;
+						break;
+					case "onlyStrict":
+						onlyStrict = true;
+						break;
+					case "async":
+						async = true;
+						break;
+					case "module":
+						module = true;
+						break;
+				}
+			}
+		}
+		ImmutableSet<String> featureSet = ImmutableSet.emptyUsingEquality();
+		Object rawFeatures = parsedYaml.get("features");
+		if (rawFeatures != null) {
+			ArrayList<String> features = (ArrayList<String>) rawFeatures;
+			for (String feature : features) {
+				featureSet = featureSet.put(feature);
+			}
+		}
+		return new Test262Info(path, negativeEnum, noStrict, onlyStrict, async, module, featureSet);
+	}
+
+	private static final class Test262Exception extends RuntimeException {
+
+		@Nonnull
+		public final String name;
+
+		public Test262Exception(@Nonnull String name, @Nonnull String message) {
+			super(message);
+			this.name = name;
+		}
+
+		public Test262Exception(@Nonnull String name, @Nonnull String message, @Nonnull Throwable caused) {
+			super(message, caused);
+			this.name = name;
+		}
+	}
+
+	private void runTest262Test(@Nonnull String source, @Nonnull Path path, @Nonnull Test262Info info, boolean strict) {
+		boolean shouldFail = info.negative == Test262Info.Test262Negative.PARSE || info.negative == Test262Info.Test262Negative.EARLY;
+		boolean xfailed = XFailHelper.isXFailed(info, shouldFail);
+		boolean failed;
+		Throwable failureReason;
+		try {
+			Program program;
+			if (info.module) {
+				program = Parser.parseModule(source);
+			} else if (strict) {
+				program = Parser.parseScript("\"use strict\";\n" + source);
+			} else {
+				program = Parser.parseScript(source);
+			}
+			ImmutableList<EarlyError> earlyErrors = EarlyErrorChecker.validate(program);
+			failed = earlyErrors.length > 0;
+			failureReason = new RuntimeException("Early Errors: \n" + earlyErrors.foldLeft((str, error) -> error.message + ", " + str, ""));
+		} catch (JsError e) {
+			failed = true;
+			failureReason = e;
+		}
+		if (xfailed && failed == shouldFail || !xfailed && failed != shouldFail) {
+			if (shouldFail != xfailed) {
+				throw new Test262Exception(info.name, "Parsed and should not have: " + path.toString());
```

Comment:
I would prefer more informative messages. For example, this message does not make it clear if the problem is that the test was xfailed and passed, or was not xfailed but failed.

I think each of the four failure cases (where `[ xfailed, shouldFail, failed ]` takes values `[ false, false, true ]`, `[ false, true, false ]`, `[ true, false, false ]`, `[ true, true, true ]`) warrants its own error message.